src/projectM-pulseaudio/qprojectM-pulseaudio.cpp

```diff
@@ -158,50 +158,58 @@ int main ( int argc, char*argv[] )
 
 std::string read_config()
 {
-
-	int n;
-
 	char num[512];
 	FILE *in;
 	FILE *out;
 
 	char* home;
-	char projectM_home[1024];
+        char* xdg_home;
+        
 	char projectM_config[1024];
+	char default_config[1024];
+
+	strcpy(default_config, PROJECTM_PREFIX);
+        strcat(default_config, CONFIG_FILE);
+	printf("Default config: %s\n", default_config);
 
-	strcpy ( projectM_config, PROJECTM_PREFIX );
-	strcpy ( projectM_config+strlen ( PROJECTM_PREFIX ), CONFIG_FILE );
-	projectM_config[strlen ( PROJECTM_PREFIX ) +strlen ( CONFIG_FILE ) ]='\0';
-	printf ( "dir:%s \n",projectM_config );
-	home=getenv ( "HOME" );
-	strcpy ( projectM_home, home );
-	strcpy ( projectM_home+strlen ( home ), "/.projectM/config.inp" );
-	projectM_home[strlen ( home ) +strlen ( "/.projectM/config.inp" ) ]='\0';
+	home=getenv("HOME");
+        xdg_home = getenv("XDG_CONFIG_HOME");
 
+	strcpy(projectM_config, home);
+        strcat(projectM_config, "/.projectM/config.inp");
 
-	if ( ( in = fopen ( projectM_home, "r" ) ) != 0 )
+	if ( ( in = fopen ( projectM_config, "r" ) ) != 0 )
 	{
-		printf ( "reading ~/.projectM/config.inp \n" );
+		printf ( "reading %s \n", projectM_config );
 		fclose ( in );
-		return std::string ( projectM_home );
+		return std::string ( projectM_config );
 	}
-	else
-	{
-		printf ( "trying to create ~/.projectM/config.inp \n" );
-
-		strcpy ( projectM_home, home );
-		strcpy ( projectM_home+strlen ( home ), "/.projectM" );
-		projectM_home[strlen ( home ) +strlen ( "/.projectM" ) ]='\0';
-		mkdir ( projectM_home, 0755 );
-
-		strcpy ( projectM_home, home );
-		strcpy ( projectM_home+strlen ( home ), "/.projectM/config.inp" );
-		projectM_home[strlen ( home ) +strlen ( "/.projectM/config.inp" ) ]='\0';
-
-		if ( ( out = fopen ( projectM_home,"w" ) ) !=0 )
-		{
-
-			if ( ( in = fopen ( projectM_config, "r" ) ) != 0 )
+	else {
+                if (xdg_home) {
```

Comment:
it looks like this still ignores the old config file if XDG home is set even if the old config file exists?
or am I misreading?
sort of hard to follow because of all the branches and lack of commenting...

Reply:
Yeah, the whole function could use some cleanup to be honest. Given my C and C++ are a bit rough I didn't want to move things around too much.

The old config file is checked for and read here at line 181.

I considered reworking the entire function to make things much easier to read but I held off because of various concerns. My initial thought had been to start using things like std::string instead of C char arrays and fixing the branching and adding comments as well.

The main thing that stopped me however was concerns over portability of using things like std::string and changing the fopen calls to something like fstream which provides other things like type safety. As I've been out of the C/C++ game for a while now, I wasn't sure about the implications of making these sort of sweeping changes.

If updating the function as a whole and using more modern libraries isn't a big concern, I can go that route and just rewrite the whole thing to clean it up to make function much more readable. I just didn't want to clobber the whole function without at least having a conversation about it.

Reply:
we're using some modern C++, C++14 I think? using fstream etc is totally fine. rewriting things is totally fine. can merge this now if you want, or can wait for you

Reply:
Why don't we go ahead and get this merged if you're satisfied with it.

That will give me some time to brush up on my C++ and follow up with a second pull request to tackle the rewrite in one go.  It should also serve as a good reference for me during refactoring.